File: ExtremeParser/Global/defineReplacer.vb
Imports System.Text.RegularExpressions

Public Class defineReplacer
    Public Shared Sub Replace(ByRef Code As String, ByVal defineName As String, ByVal defineReplace As String, isMacro As Boolean)
        'First start building the regex for the find.
        Dim findRegex As String = Regex.Escape(defineName)

        'Setup find regex.
        findRegex = Regex.Replace(findRegex, "%([0-9])", "(.*)")
        findRegex = "(?<!#.*)\b" + findRegex
        If isMacro = False Then findRegex += "\b"

        'Setup replace regex.
        Dim replaceRegex As String = defineReplace
        If replaceRegex.Contains("%0") Then
            For i As Integer = 0 To 9
                replaceRegex = replaceRegex.Replace("%" + i.ToString, "$" + (i + 1).ToString)
            Next
        Else
            For i As Integer = 1 To 9
                replaceRegex = replaceRegex.Replace("%" + i.ToString, "$" + i.ToString)
            Next
        End If

        Code = Regex.Replace(Code, findRegex, replaceRegex)
    End Sub
End Class

File: ExtremeParser/Global/DefineReplacer.cs
using System;
using System.Text.RegularExpressions;

namespace ExtremeParser.Global
{
    public class DefineReplacer
    {
        public static void Replace(ref string code, string defineName, string defineReplace, bool isMacro)
        {
            //First start building the regex for the find.
            string findRegex = Convert.ToString(Regex.Escape(defineName));
		
            //Setup find regex.
            findRegex = Convert.ToString(Regex.Replace(findRegex, "%([0-9])", "(.*)"));
            findRegex = "(?<!#.*)\\b" + findRegex;
            if (isMacro == false)
            {
                findRegex += "\\b";
            }
		
            //Setup replace regex.
            string replaceRegex = defineReplace;
            if (replaceRegex.Contains("%0"))
            {
                for (int i = 0; i <= 9; i++)
                {
                    replaceRegex = replaceRegex.Replace("%" + i, "$" + (i + 1));
                }
            }
            else
            {
                for (int i = 1; i <= 9; i++)
                {
                    replaceRegex = replaceRegex.Replace("%" + i, "$" + i);
                }
            }
		
            code = Convert.ToString(Regex.Replace(code, findRegex, replaceRegex));
        }
    }
}
